### DISCUSSION TITLE
Uploading files in python not working

### DISCUSSION THREAD

**Discussion by IttsKK** [etsy/open-api#967] [Q&A] [answered]
Whenever I try to upload a file to a listing using the api from a python client, I get multiple errors:

When using `multipart/form-data` I get `{"error":"An error ocurred while uploading your file, please try again."}` and when using `application/x-www-form-urlencoded` I get `{"error":"Either a valid listing_file_id or file must be provided."}`

Here's my code for reference:

```
import requests
import os
from dotenv import load_dotenv
load_dotenv()

api_key = os.getenv('etsy_keystring')
access_token = 'Bearer ' + os.getenv('etsy_access_token')
shop_id = os.getenv('shop_id')
listing_id = '1234567890'

file_url = "https://openapi.etsy.com/v3/application/shops/" + str(shop_id) + "/listings/" + str(listing_id) + "/files"

headers = {
    "Content-Type": "multipart/form-data",
    "x-api-key": api_key,
    "Authorization": access_token
}

with open('/path/to/image', 'rb') as f:
    file_data = f.read()
    file_name = os.path.basename('/path/to/image')

data = {
    'name': file_name
}

files = {
    'file': (file_name, file_data),
}

response = requests.post(file_url, headers=headers, data=data, files=files)

print(response.text)
```

Thanks for any help!

Edit: I have no trouble adding images to a listing, the problem arises when trying to add a file to a digital listing

**Comment by IanDanielM** [comment]
First, remove the `"Content-Type"` header with the value `"multipart/form-data"`. I think they should resolve that issue from their end.

```
files={"image": open(r"filepath/filename","rb"  }
headers = {'x-api-key': 'xxxxxxxxxxxxxx',
           'Authorization': auth
                   }
r=requests.post(url,headers=headers,files=files)
```

this should go through

**Reply by IttsKK** [reply]
This is what I do for adding an image to listings, but it just doesn't work when adding a file.

**Reply by stackiflowi** [reply]
To me this is a working way:

`
data = {'rank': 5}

files={'image': [image_name, open(image_path, 'rb'), 'multipart/form-data']}

r = requests.post(url, headers=header, data=data, files=files)
`

I hope it helps. Works for me for digital listings as well.

**Reply by IttsKK** [reply]
Strange, when I use this method I get the message
"error":"A valid name must be provided with a new file."
also I'm pretty sure the file should be named "file" not "image".

**Reply by DevDenzo** [reply]
Did you figure this out? If so I would appreciate if you could share your solution.

**Comment by IttsKK** [comment, marked as answer]
Found a solution, thanks to @stackiflowi:

data = {'rank': 5, 'name': 'file_name'}

files={'file': ["image", open(file_path, 'rb'), 'multipart/form-data']}

r = requests.post(file_url, headers=headers, data=data, files=files)

**Reply by sebing** [reply]
not working :(

**Reply by DevDenzo** [reply]
This isnt working for me either I am still also getting:

{"error":"Either a valid listing_file_id or file must be provided."}

Has anyone solved this?

**Reply by olehsharov** [reply]
@DevDenzo  @sebing  
Try removing 'multipart/form-data'.
Did work for my js code and if it does - docs are misleading.

**Comment by sebing** [comment]
What am I doing wrong?
```
data = {'rank': 5, 'name': 'file_name'}
files = {'file': ["image", open('1.jpg', 'rb'), 'multipart/form-data']}
requests.post(url, headers=headers, data=data, files=files).json()

 {'error': 'Either a valid listing_file_id or file must be provided.'}
```

**Reply by stackiflowi** [reply]
@sebing The code looks ok if there's no misspell in it.

Maybe You could check the "1.jpg". Is it in the right place? Maybe use full path. Check permissions.

Is there're any other image with the same rank? I guess in this case the error message should be different, but not sure.

url?

**Reply by sebing** [reply]
@stackiflowi if the file is not there i am getting a different error
same with the uri. 
 {'error': 'Either a valid listing_file_id or file must be provided.'} seems those two are not the issue.

**Reply by sebing** [reply]
Resolved the issue
```
data = {'rank': 5, 'name': 'file_name'}
```
to 
```
data = {'rank': 5, 'name': 'file_name.jpg'}
```

**Reply by heaven7** [reply]
@sebing can you please show your full working code?

**Reply by DevDenzo** [reply]
Even after adding .jpg this still doesn't work. Please could someone show full code solution for this issue

**Reply by olehsharov** [reply]
@DevDenzo  @sebing  
Try removing 'multipart/form-data'.
Did work for my js code and if it does - docs are misleading.
